stan/math/prim/scal/fun/grad_reg_inc_gamma.hpp

```diff
@@ -20,20 +21,46 @@ namespace stan {
       using std::fabs;
       using std::log;
 
-      T S = 0;
-      T s = 1;
       T l = log(z);
-      int k = 0;
-      T delta = s / (a * a);
-      while (fabs(delta) > precision) {
-        S += delta;
-        ++k;
-        s *= - z / k;
-        delta = s / ((k + a) * (k + a));
-        if (isinf(delta))
-          throw domain_error("gradRegIncGamma not converging");
+      if (z >= a && z >= (T)8) {
+        // large values of z, compute gradient based on the asymptotic expansion
+        // http://dlmf.nist.gov/8.11#E2
+        T S = 0;
+        T fac = a-1;  // falling_factorial(a-1, k)
+        T dfac = 1;   // d/da[falling_factorial(a-1, k)]
+        T zpow = z;   // z ** k
+        T delta = dfac / zpow;
+
+        for (int k = 1; k < 10; ++k) {
+          S += delta;
+
+          zpow *= z;
+          dfac = (a-1-k)*dfac + fac;
+          fac *= a-1-k;
+          delta = dfac / zpow;
+
```

Comment:
For efficiency,

* store `a - 1 - k`, which is `fac - k` in a local and reuse.  

Also, please add spaces around operators (subtraction, addition, multiplication, etc.)